### PR TITLE
Create multi-stage RISCV-32 build

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
This replaces the Kokoro pipeline and demonstrates a slightly more
complex multi-stage build. The host tools build here should be reusable
for any other job that needs the host tools, though note that these
ones are built with asserts enabled and we'll probably want another job
that builds with them disabled (e.g. for benchmarking).

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
Uuuuuh... I don't really understand how this could've improved compilation times and regressed run times. Assertions are turned off: https://buildkite.com/iree/iree-benchmark-linux/builds/810#01821d4a-4f74-45b1-ba61-9cde3ac02ef6/43-75

Linking with lld couldn't cause this, could it??

Wait, hang on. The pipeline actually rebuilds the benchmark tools separately from the host tools. So these regressions look spurious.

**Comment by GMNGeoffrey** [comment]
I'm guessing there has been some improvement on mainline in the meantime. I guess I can merge from main and find out. And then rerun all the tests...

**Comment by pzread** [comment]
The compilation time is measured when building the benchmark suites (vmfb): https://buildkite.com/iree/iree-benchmark-linux/builds/810#01821d4a-4f80-43ad-aa1b-1795a1d2108f

It is not fine tuned yet (and the build machine is just a normal builder), so they might be just the noise or improvement from other places. The latency regression also looks random, might be the noise.

File: build_tools/cmake/build_host_and_riscv.sh [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+# Cross-compile the IREE project towards RISCV with CMake. Designed for CI,
+# but can be run manually. This uses previously cached build results and does
+# not clear build directories.
+#
+# Host binaries (e.g. compiler tools) will be built and installed in build-host/
+# RISCV binaries (e.g. tests) will be built in build-riscv/.
+
+set -x
+set -e
```
Comment (pzread) [inline build_tools/cmake/build_host_and_riscv.sh:17]: Any reason not uses set -euo -pipefail?

File: build_tools/cmake/build_host_and_riscv.sh [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+"${CMAKE_BIN?}" -G Ninja -B "${BUILD_HOST_DIR?}" \
+  -DCMAKE_INSTALL_PREFIX="${BUILD_HOST_DIR?}/install" \
+  -DCMAKE_C_COMPILER="${CC:-clang}" \
+  -DCMAKE_CXX_COMPILER="${CXX:-clang++}" \
+  -DIREE_ENABLE_ASSERTIONS=ON \
+  -DIREE_BUILD_COMPILER=ON \
+  -DIREE_BUILD_TESTS=OFF \
+  -DIREE_BUILD_SAMPLES=OFF \
+  "${ROOT_DIR?}"
+"${CMAKE_BIN?}" --build "${BUILD_HOST_DIR?}" --target install -- -k 0
```
Comment (pzread) [inline build_tools/cmake/build_host_and_riscv.sh:49]: Can we directly invoke `build_host_tools.sh` here?

File: build_tools/cmake/build_host_and_riscv.sh [hunk 3]
```diff
@@ -0,0 +1,94 @@
+#!/bin/bash
```
Comment (pzread) [inline build_tools/cmake/build_host_and_riscv.sh:1]: Realized this file is renamed from `build_riscv.sh`. Feel free to ignore those comments.

**Review by pzread — APPROVED** [review]

**Comment by hcindyl** [comment]
nit: the title should be both RISCV-64 and RISCV-32 builds.

File: build_tools/cmake/build_host_and_riscv.sh (hunk 3 above)
Comment (GMNGeoffrey) [reply build_tools/cmake/build_host_and_riscv.sh:1]: Yeah I want to just delete this file shortly :-)

**Comment by GMNGeoffrey** [comment]
> nit: the title should be both RISCV-64 and RISCV-32 builds.

You mean the PR title? I didn't make a RISCV-64 build though. I just renamed its build script

**Comment by hcindyl** [comment]
> > nit: the title should be both RISCV-64 and RISCV-32 builds.
> 
> You mean the PR title? I didn't make a RISCV-64 build though. I just renamed its build script

ahh. I see. sorry I misread. Are we going to have a RISCV-64 github action added later?

**Comment by GMNGeoffrey** [comment]
> > > nit: the title should be both RISCV-64 and RISCV-32 builds.
> > 
> > 
> > You mean the PR title? I didn't make a RISCV-64 build though. I just renamed its build script
> 
> ahh. I see. sorry I misread. Are we going to have a RISCV-64 github action added later?

Yup :slightly_smiling_face:  https://github.com/iree-org/iree/issues/9855

**Comment by GMNGeoffrey** [comment]
Forgot to tag originally, but part of https://github.com/iree-org/iree/issues/9855

**Pull request merged**